### PR TITLE
Replace the empty container with a local volume

### PR DESCRIPTION
**Pull request by richardpen**
### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: 
yes
### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/api/task/task.go [hunk 1]
```diff
@@ -202,55 +206,53 @@ func (task *Task) PostUnmarshalTask(cfg *config.Config,
...
-			Command:             []string{emptyvolume.Command}, // Command required, but this only gets created so N/A
-			MountPoints:         mountPoints,
-			Essential:           false,
-			Type:                apicontainer.ContainerEmptyHostVolume,
-			DesiredStatusUnsafe: apicontainer.ContainerRunning,
-		}
-		task.Containers = append(task.Containers, sourceContainer)
+	// if we have required local volumes, create one with default local drive
+	for _, volumeName := range requiredLocalVolumes {
+		vol, _ := task.HostVolumeByName(volumeName)
```
Comment (tehshwann) [inline agent/api/task/task.go]: Im not sure how this code interacts with the Windows disk manager, so I dont have a good solution, but I am concerned that we arent using a unique identifier here.

What are we doing to ensure we are choosing the right volume and arent conflicting spaces between tasks?

File: agent/api/task/task.go [hunk 2]
```diff
@@ -202,55 +206,53 @@ func (task *Task) PostUnmarshalTask(cfg *config.Config,
...
+		// BUG(samuelkarp) On Windows, volumes with names that differ only by case will collide
+		scope := taskresourcevolume.TaskScope
+		autoprovision := false // must be false if scope = task
+		localDriver := "local"
+		localVolume := taskresourcevolume.NewVolumeResource(volumeName, vol.SourcePath(), scope, autoprovision, localDriver, make(map[string]string), make(map[string]string), dockerClient)
```
Comment (tehshwann) [inline agent/api/task/task.go]: Does this driver work with Windows? If so, how are we excersizing this code path on Windows in testing?

File: agent/api/task/task.go [hunk 3]
```diff
@@ -202,55 +206,53 @@ func (task *Task) PostUnmarshalTask(cfg *config.Config,
...
+		task.AddResource(volumeName, localVolume)
 	}
 }
 
+func (task *Task) taskScopedVolumeName(name string) string {
```
Comment (tehshwann) [inline agent/api/task/task.go]: does this need to be OS aware?

File: agent/api/task/task.go [hunk 4]
```diff
@@ -347,7 +349,7 @@ func (task *Task) ContainerByName(name string) (*apicontainer.Container, bool) {
 
 // HostVolumeByName returns the task Volume for the given a volume name in that
 // task. The second return value indicates the presence of that volume
-func (task *Task) HostVolumeByName(name string) (HostVolume, bool) {
+func (task *Task) HostVolumeByName(name string) (taskresourcevolume.HostVolume, bool) {
```
Comment (tehshwann) [inline agent/api/task/task.go]: Is the hostvolume something the user is volunteering to be used by the task? Or are we making that assumption for the user?

Concern:
It seems risky to make any assumptions related to disk space usage since there are limited tools, telemetry, and "rails" to manage disk utilization on the host to prevent disk space related issues.

File: agent/engine/engine_windows_integ_test.go [hunk 5]
```diff
@@ -68,18 +72,14 @@ func createTestHostVolumeMountTask(tmpPath string) *apitask.Task {
...
 	testTask.Containers[0].Image = testVolumeImage
-	testTask.Containers[0].MountPoints = []apicontainer.MountPoint{{ContainerPath: "C:/empty", SourceVolume: "test-tmp"}}
-	testTask.Containers[0].Command = []string{`While($true){ if (Test-Path C:\empty\file) { exit 42 } }`}
-	testTask.Containers = append(testTask.Containers, createTestContainer())
-	testTask.Containers[1].Name = "test2"
-	testTask.Containers[1].Image = testVolumeImage
-	testTask.Containers[1].MountPoints = []apicontainer.MountPoint{{ContainerPath: "C:/alsoempty/", SourceVolume: "test-tmp"}}
-	testTask.Containers[1].Command = []string{`New-Item -Path C:\alsoempty\file`}
-	testTask.Containers[1].Essential = false
+	testTask.Containers[0].Command = []string{`echo -n empty-data-volume | Out-File -FilePath C:\host\tmp\hello-from-container -Encoding ascii`}
```
Comment (tehshwann) [inline agent/engine/engine_windows_integ_test.go]: Im not sure what "echo -n" is... but echo is a cmd.exe that is aliased in PowerShell to "Write-Output" which doesnt have a "-n" arg.

This would be better powershell assuming you just want to print 'empty-data-volume':

`testTask.Containers[0].Command = []string{'Write-Output "empty-data-volume" | Out-File -FilePath C:\host\tmp\hello-from-container -Encoding ascii'}`

File: agent/engine/engine_windows_integ_test.go [hunk 6]
```diff
@@ -68,18 +72,14 @@ func createTestHostVolumeMountTask(tmpPath string) *apitask.Task {
...
+	testTask.Containers[0].MountPoints = []apicontainer.MountPoint{{ContainerPath: "C:\\host\\tmp", SourceVolume: "test-tmp"}}
```
Comment (tehshwann) [inline agent/engine/engine_windows_integ_test.go]: Also, "C:\host\tmp\" is fine for testing use. But a more traditional path for powershell/windows would be "$ENV:TEMP" for the current user's temp dir. I would consider using this so that running the tests dont create unnecessary and unconventional directories on the root volume

File: agent/engine/engine_windows_integ_test.go [hunk 7]
```diff
@@ -127,6 +127,30 @@ func getContainerIP(client *docker.Client, id string) (string, error) {
...
+	verifyContainerStoppedStateChange(t, taskEngine)
+	verifyTaskIsStopped(stateChangeEvents, testTask)
+
+	assert.NotNil(t, testTask.Containers[0].GetKnownExitCode(), "No exit code found")
+	assert.Equal(t, 0, *testTask.Containers[0].GetKnownExitCode(), "Wrong exit code")
+
+	assert.Contains(t, testTask.Volumes[0].Volume.SourcePath(), "C:\\ProgramData\\docker\\volumes")
+	data, err := ioutil.ReadFile(filepath.Join(testTask.Volumes[0].Volume.SourcePath(), "hello-from-container"))
+	assert.Nil(t, err, "Unexpected error")
+	assert.Equal(t, "empty-data-volume", strings.TrimSpace(string(data)), "Incorrect file contents")
```
Comment (tehshwann) [inline agent/engine/engine_windows_integ_test.go]: Im not sure how the Go libs work with ntfs here.. but will this opperate correctly if there is an EOF character?

File: agent/engine/task_manager_test.go [hunk 8]
```diff
@@ -1622,6 +1623,255 @@ func TestApplyResourceStateFailures(t *testing.T) {
...
+				Task: &apitask.Task{},
+			}
+			transition := mtask.resourceNextState(&res)
+			assert.Equal(t, tc.NextState, transition.nextState)
+			assert.Equal(t, tc.ActionRequired, transition.actionRequired)
+		})
+	}
+}
+
+func TestStartVolumeResourceTransitionsHappyPath(t *testing.T) {
```
Comment (tehshwann) [inline agent/engine/task_manager_test.go]: Should we be testing non-empty/non-happy path transitions? For example, if there are timeouts waiting for transition, or Docker has an issue handling the volume?

File: agent/taskresource/volume/dockervolume.go [hunk 9]
```diff
@@ -53,15 +71,17 @@ type DockerVolumeConfig struct {
 
 // NewVolumeResource returns a docker volume wrapper object
 func NewVolumeResource(name string,
+	dockerVolumeName string,
```
Comment (tehshwann) [inline agent/taskresource/volume/dockervolume.go]: Are these guaranteed to be unique? Does that matter to us?

File: agent/taskresource/volume/dockervolume.go [hunk 10]
```diff
@@ -71,40 +91,124 @@ func NewVolumeResource(name string,
...
+// SteadyState returns the transition state of the resource defined as "ready"
+func (vol *VolumeResource) SteadyState() taskresource.ResourceStatus {
+	return taskresource.ResourceStatus(VolumeCreated)
+}
+
+// ApplyTransition calls the function required to move to the specified status
+func (vol *VolumeResource) ApplyTransition(nextState taskresource.ResourceStatus) error {
+	transitionFunc, ok := vol.resourceStatusToTransitionFunction[nextState]
+	if !ok {
+		seelog.Errorf("Volume Resource [%s]: unsupported desired state transition: %s",
```
Comment (tehshwann) [inline agent/taskresource/volume/dockervolume.go]: Is this the right level of detail for failing to transition the state of a critical task resource?

Comment (tehshwann) [reply agent/taskresource/volume/dockervolume.go]: How would failures here be debugged if, for example, docker was having an issue with the volume that bubbled up here and resulted in a timeout for a state transition?

File: agent/api/task/task.go [hunk 11]
```diff
@@ -202,55 +206,53 @@ func (task *Task) PostUnmarshalTask(cfg *config.Config,
...
 	task.initializeCredentialsEndpoint(credentialsManager)
 	task.addNetworkResourceProvisioningDependency(cfg)
 	return nil
 }
 
-func (task *Task) initializeEmptyVolumes() {
-	requiredEmptyVolumes := []string{}
+func (task *Task) initializeLocalVolumes(dockerClient dockerapi.DockerClient) {
+	requiredLocalVolumes := []string{}
+	reInvalidChars := regexp.MustCompile("[^A-Za-z0-9-]+")
```
Comment (aaithal) [inline agent/api/task/task.go]: Why is this needed? Also, this should be a global var if needed.

File: agent/api/task/task.go (hunk 1 above)
Comment (aaithal) [inline agent/api/task/task.go]: Why are you ignoring the returned boolean here?

File: agent/api/task/task.go [hunk 12]
```diff
@@ -202,55 +206,53 @@ func (task *Task) PostUnmarshalTask(cfg *config.Config,
...
-			Type:                apicontainer.ContainerEmptyHostVolume,
-			DesiredStatusUnsafe: apicontainer.ContainerRunning,
-		}
-		task.Containers = append(task.Containers, sourceContainer)
+	// if we have required local volumes, create one with default local drive
+	for _, volumeName := range requiredLocalVolumes {
+		vol, _ := task.HostVolumeByName(volumeName)
+		// BUG(samuelkarp) On Windows, volumes with names that differ only by case will collide
+		scope := taskresourcevolume.TaskScope
+		autoprovision := false // must be false if scope = task
```
Comment (aaithal) [inline agent/api/task/task.go]: %s/autoprovision/autoProvision/g

File: agent/api/task/task.go [hunk 13]
```diff
@@ -202,55 +206,53 @@ func (task *Task) PostUnmarshalTask(cfg *config.Config,
...
+		localDriver := "local"
```
Comment (aaithal) [inline agent/api/task/task.go]: Please make this a const

File: agent/api/task/task.go (hunk 2 above)
Comment (aaithal) [inline agent/api/task/task.go]: please split this into multiple lines

File: agent/taskresource/volume/dockervolume.go [hunk 14]
```diff
@@ -15,23 +15,41 @@ package volume
...
+	"github.com/pkg/errors"
 
 	"encoding/json"
 	"sync"
 	"time"
 )
 
-// VolumeResource represents volume resource
+const (
+	//TaskScope indicates that the volume is created and deleted with task
```
Comment (aaithal) [inline agent/taskresource/volume/dockervolume.go]: there needs to be a space between '//' and the comment. Please fix this everywhere

File: agent/taskresource/volume/dockervolume.go [hunk 15]
```diff
@@ -15,23 +15,41 @@ package volume
...
-	knownStatusUnsafe   VolumeStatus
-	client              dockerapi.DockerClient
+	desiredStatusUnsafe taskresource.ResourceStatus
+	knownStatusUnsafe   taskresource.ResourceStatus
+	// appliedStatus is the status that has been "applied" (e.g., we've called some
+	// operation such as 'Create' on the resource) but we don't yet know that the
+	// application was successful, which may then change the known status. This is
+	// used while progressing resource states in progressTask() of task manager
+	appliedStatus                      taskresource.ResourceStatus
+	resourceStatusToTransitionFunction map[taskresource.ResourceStatus]func() error
```
Comment (aaithal) [inline agent/taskresource/volume/dockervolume.go]: `statusToTransitions` is a smaller and friendlier name for this.

File: agent/taskresource/volume/dockervolume.go [hunk 16]
```diff
@@ -71,40 +91,124 @@ func NewVolumeResource(name string,
 		},
 		client: client,
 	}
+	v.initializeResourceStatusToTransitionFunction(scope)
+	return v
+}
+
+func (vol *VolumeResource) initializeResourceStatusToTransitionFunction(scope string) {
```
Comment (aaithal) [inline agent/taskresource/volume/dockervolume.go]: Related to my previous comment, `initStatusToTransitions` is a smaller and friendlier name for this.

File: agent/taskresource/volume/dockervolume.go (hunk 10 above)
Comment (aaithal) [inline agent/taskresource/volume/dockervolume.go]: You don't need this log. It's already being logged in the engine

File: agent/api/task/task.go [hunk 17]
```diff
@@ -202,55 +205,54 @@ func (task *Task) PostUnmarshalTask(cfg *config.Config,
...
 		for _, mountPoint := range container.MountPoints {
 			vol, ok := task.HostVolumeByName(mountPoint.SourceVolume)
 			if !ok {
 				continue
 			}
-			if _, ok := vol.(*EmptyHostVolume); ok {
-				container.BuildContainerDependency(emptyHostVolumeName, apicontainer.ContainerRunning, apicontainer.ContainerCreated)
-				requiredEmptyVolumes = append(requiredEmptyVolumes, mountPoint.SourceVolume)
+			if localVolume, ok := vol.(*taskresourcevolume.LocalVolume); ok {
+				localVolume.HostPath = task.taskScopedVolumeName(mountPoint.SourceVolume)
```
Comment (aaithal) [inline agent/api/task/task.go]: `task.taskScopedVolumeName` stutters. How about renaming it to just `VolumeName`?

File: agent/api/task/task.go [hunk 18]
```diff
@@ -202,55 +205,54 @@ func (task *Task) PostUnmarshalTask(cfg *config.Config,
...
-			Type:                apicontainer.ContainerEmptyHostVolume,
-			DesiredStatusUnsafe: apicontainer.ContainerRunning,
-		}
-		task.Containers = append(task.Containers, sourceContainer)
+	// if we have required local volumes, create one with default local drive
+	for _, volumeName := range requiredLocalVolumes {
+		vol, _ := task.HostVolumeByName(volumeName)
+		// BUG(samuelkarp) On Windows, volumes with names that differ only by case will collide
+		scope := taskresourcevolume.TaskScope
+		autoProvision := false // must be false if scope = task
```
Comment (aaithal) [inline agent/api/task/task.go]: If this is a prereq, it should be enforced in the `NewVolumeResource`  method, where you need to check for that condition and return an error.

File: agent/taskresource/volume/dockervolume.go [hunk 19]
```diff
@@ -15,22 +15,42 @@ package volume
...
+	// DockerLocalVolumeDriver is the name of the docker default volume driver
+	DockerLocalVolumeDriver = "local"
+)
+
+// VolumeResources represents volume resource
 type VolumeResource struct {
-	// Name is the name of docker volume
+	// Name is the name of the docker volume
 	Name string
+	// dockerVolumeName is internal docker name for this volume.
```
Comment (aaithal) [inline agent/taskresource/volume/dockervolume.go]: nit: DockerVolumeName

File: agent/taskresource/volume/dockervolume.go [hunk 20]
```diff
@@ -15,22 +15,42 @@ package volume
...
 	createdAtUnsafe     time.Time
-	desiredStatusUnsafe VolumeStatus
-	knownStatusUnsafe   VolumeStatus
+	desiredStatusUnsafe taskresource.ResourceStatus
+	knownStatusUnsafe   taskresource.ResourceStatus
+	// appliedStatus is the status that has been "applied" (e.g., we've called some
+	// operation such as 'Create' on the resource) but we don't yet know that the
+	// application was successful, which may then change the known status. This is
+	// used while progressing resource states in progressTask() of task manager
+	appliedStatus       taskresource.ResourceStatus
```
Comment (aaithal) [inline agent/taskresource/volume/dockervolume.go]: Should this be `appliedStatusUnsafe` as well?

File: agent/taskresource/volume/dockervolume.go [hunk 21]
```diff
@@ -71,40 +93,118 @@ func NewVolumeResource(name string,
...
+// SteadyState returns the transition state of the resource defined as "ready"
+func (vol *VolumeResource) SteadyState() taskresource.ResourceStatus {
+	return taskresource.ResourceStatus(VolumeCreated)
+}
+
+// ApplyTransition calls the function required to move to the specified status
+func (vol *VolumeResource) ApplyTransition(nextState taskresource.ResourceStatus) error {
+	transitionFunc, ok := vol.statusToTransitions[nextState]
+	if !ok {
+		return errors.Errorf("resource [%s]: transition to %s impossible", vol.Name,
```
Comment (aaithal) [inline agent/taskresource/volume/dockervolume.go]: Why not begin the error string with 'volume' instead of 'resource' ?

File: agent/taskresource/volume/dockervolume.go [hunk 22]
```diff
@@ -162,6 +267,12 @@ func (vol *VolumeResource) Create() error {
 
 // Cleanup performs resource cleanup
 func (vol *VolumeResource) Cleanup() error {
+	// Enable volume clean up if it's task scoped
+	if vol.VolumeConfig.Scope != TaskScope {
+		seelog.Debugf("Volume is shared, not removing", vol.Name)
```
Comment (aaithal) [inline agent/taskresource/volume/dockervolume.go]: You're missing a %s here

File: agent/taskresource/volume/volumetypes.go [hunk 23]
```diff
@@ -30,12 +30,14 @@ func (fs *FSHostVolume) SourcePath() string {
 	return fs.FSSourcePath
 }
 
-// EmptyHostVolume represents a volume without a specified host path
-type EmptyHostVolume struct {
+// LocalVolume represents a volume without a specified host path
+// This is essentially DockerVolume with only the name specified; however,
+// for backward compatibility we can't directly map to DockerVolume.
+type LocalVolume struct {
```
[the remark this answers is not in the document]
Comment (aaithal) [reply agent/taskresource/volume/volumetypes.go]: This is still not addressed.

File: agent/taskresource/volume/dockervolume.go [hunk 24]
```diff
@@ -15,22 +15,42 @@ package volume
...
 
 	"encoding/json"
 	"sync"
 	"time"
 )
 
-// VolumeResource represents volume resource
+const (
+	// TaskScope indicates that the volume is created and deleted with task
+	TaskScope = "task"
```
Comment (aaithal) [inline agent/taskresource/volume/dockervolume.go]: It seems like both `TaskScope ` and `SharedScope` should be enums. wdyt?

Comment (richardpen) [reply agent/taskresource/volume/dockervolume.go]: yes, I agree `enums` makes it looks better, but using constants here makes it simple to avoid conversion between string and enums, as this is what backend send to us. I think they works the same as enums is implemented based on constants, so I will keep this as it is unless you strongly want to change this.

File: agent/taskresource/volume/dockervolume.go [hunk 25]
```diff
@@ -76,40 +102,124 @@ func NewVolumeResource(name string,
...
+	vol.client = resourceFields.DockerClient
+	vol.initStatusToTransitions()
+}
+
+func (vol *VolumeResource) initStatusToTransitions() {
+	statusToTransitions := map[resourcestatus.ResourceStatus]func() error{
+		resourcestatus.ResourceStatus(VolumeCreated): vol.Create,
+	}
+
+	statusToTransitions[resourcestatus.ResourceStatus(VolumeRemoved)] = vol.Cleanup
```
Comment (sharanyad) [inline agent/taskresource/volume/dockervolume.go]: `Cleanup` need not be added to the status to transitions map since it is not called during `progressTask`. Resources cleanup happen here: 
https://github.com/aws/amazon-ecs-agent/blob/8dae73394282df16570d41c067fa364652db4e31/agent/engine/docker_task_engine.go#L464

File: agent/taskresource/volume/dockervolume.go [hunk 26]
```diff
@@ -15,24 +15,44 @@ package volume
...
+	// TaskScope indicates that the volume is created and deleted with task
+	TaskScope = "task"
+	// SharedScope indicates that the volume's lifecycle is outside the scope of task
+	SharedScope = "shared"
+	// DockerLocalVolumeDriver is the name of the docker default volume driver
+	DockerLocalVolumeDriver = "local"
 )
 
-// VolumeResource represents volume resource
+// VolumeResources represents volume resource
```
Comment (sharanyad) [inline agent/taskresource/volume/dockervolume.go]: nit: `VolumeResource`

File: agent/engine/docker_task_engine.go [hunk 27]
```diff
@@ -323,11 +323,6 @@ func updateContainerMetadata(metadata *dockerapi.DockerContainerMetadata, contai
 		container.SetLabels(metadata.Labels)
 	}
 
-	// Update Volume
-	if metadata.Volumes != nil {
-		task.UpdateMountPoints(container, metadata.Volumes)
```
[the remark this answers is not in the document]
Comment (richardpen) [reply agent/engine/docker_task_engine.go]: This was previously used to update the mount point from the empty container, so removing this after the empty container was removed.

File: agent/taskresource/volume/dockervolume.go (hunk 25 above)
Comment (richardpen) [reply agent/taskresource/volume/dockervolume.go]: Yes, but it shouldn't hurt to add it here. As the `REMOVED` should match to the `Cleanup` in the transition function.

File: agent/api/task/task.go (hunk 4 above)
Comment (yunhee-l) [reply agent/api/task/task.go]: from moby issue #21829 (thought it might be relevant):
What is the volume size of the default volume?

Volumes created through the default ("local") volume driver don't have a size limit; they're basically just directories created inside /var/lib/docker/volumes. If you want to put a size constraint on volumes, you need to use a volume-plugin that supports this. Some plugins can be found in the documentation.

File: agent/engine/task_manager_test.go [hunk 28]
```diff
@@ -1622,6 +1624,254 @@ func TestApplyResourceStateFailures(t *testing.T) {
 	}
 }
 
+///////
```
Comment (yunhee-l) [inline agent/engine/task_manager_test.go]: nit: I had added this to mark on the file where new tests started, remove it?

File: agent/taskresource/volume/dockervolume.go (hunk 9 above)
Comment (yunhee-l) [reply agent/taskresource/volume/dockervolume.go]: Docker volume name for replacement "empty volume" are guaranteed to be unique (because we create it to force the scope to the running task). The Name for shared volume isn't, because it's used to refer to existing volume.

File: agent/api/task/task.go [hunk 29]
```diff
@@ -202,53 +206,59 @@ func (task *Task) PostUnmarshalTask(cfg *config.Config,
...
+	}
 	task.initializeCredentialsEndpoint(credentialsManager)
 	task.addNetworkResourceProvisioningDependency(cfg)
 	return nil
 }
 
-func (task *Task) initializeEmptyVolumes() {
-	requiredEmptyVolumes := []string{}
+func (task *Task) initializeDockerLocalVolumes(dockerClient dockerapi.DockerClient, ctx context.Context) error {
+	requiredLocalVolumes := []string{}
```
Comment (aaithal) [inline agent/api/task/task.go]: nit: `var requiredLocalVolumes []string` is the idiomatic way of doing this i think

File: agent/taskresource/volume/dockervolume.go [hunk 30]
```diff
@@ -56,16 +76,22 @@ type DockerVolumeConfig struct {
...
 func NewVolumeResource(name string,
+	dockerVolumeName string,
 	scope string,
 	autoprovision bool,
 	driver string,
 	driverOptions map[string]string,
 	labels map[string]string,
 	client dockerapi.DockerClient,
-	ctx context.Context) *VolumeResource {
+	ctx context.Context) (*VolumeResource, error) {
```
Comment (aaithal) [inline agent/taskresource/volume/dockervolume.go]: `ctx` should be the first param to this method

File: agent/taskresource/volume/dockervolume.go [hunk 31]
```diff
@@ -56,16 +76,22 @@ type DockerVolumeConfig struct {
...
 
-	return &VolumeResource{
-		Name: name,
+	if scope == TaskScope && autoprovision {
+		return nil, errors.New("Volume Resource creation: task scoped volume could not be autoprovisioned")
```
Comment (aaithal) [inline agent/taskresource/volume/dockervolume.go]: error string doesn't follow conventions here.

File: agent/taskresource/volume/dockervolume_test.go [hunk 32]
```diff
@@ -166,6 +221,46 @@ func TestUnmarshall(t *testing.T) {
...
+}
+
+func TestNewVolumeResource(t *testing.T) {
+	testCases := []struct {
+		scope         string
+		autoprovision bool
+		fail          bool
+	}{
+		{
+			"task",
```
Comment (aaithal) [inline agent/taskresource/volume/dockervolume_test.go]: please add a line preceding each of these testcases with a description. you can also add a `name string` field to the struct and use that as well.

File: agent/taskresource/volume/dockervolume.go (hunk 25 above)
Comment (sharanyad) [reply agent/taskresource/volume/dockervolume.go]: Since it is not being used anywhere, it would be preferable to remove this.

File: agent/app/agent_unix.go [hunk 33]
```diff
@@ -192,8 +192,10 @@ func contains(capabilities []string, capability string) bool {
 // object
 func (agent *ecsAgent) initializeResourceFields() {
 	agent.resourceFields = &taskresource.ResourceFields{
-		Control: cgroup.New(),
-		IOUtil:  ioutilwrapper.NewIOUtil(),
+		Control:      cgroup.New(),
```
Comment (sharanyad) [inline agent/app/agent_unix.go]: Is it a concern that we initialize these fields irrespective of whether resources like cgroup/volumes are enabled or not?

Comment (richardpen) [reply agent/app/agent_unix.go]: It's not a big concern to me, as it's just creating the object. This avoids an additional check, if you feel strongly about this, I can change it.

File: agent/taskresource/volume/dockervolume_test.go [hunk 34]
```diff
@@ -113,17 +115,66 @@ func TestCleanupError(t *testing.T) {
...
+	volume.ApplyTransition(resourcestatus.ResourceStatus(VolumeCreated))
+}
+
+func TestApplyTransitionForSharedScopeVolume(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	mockClient := mock_dockerapi.NewMockDockerClient(ctrl)
+
+	name := "volumeName"
+	scope := "shared"
```
Comment (sharanyad) [inline agent/taskresource/volume/dockervolume_test.go]: We do not cleanup shared volumes. Do we still expect `RemoveVolume` call here?

File: agent/taskresource/volume/dockervolume.go [hunk 35]
```diff
@@ -146,12 +255,17 @@ func (vol *VolumeResource) GetMountPoint() string {
...
 }
 
+// SourcePath is fulfilling the HostVolume interface
+func (vol *VolumeResource) SourcePath() string {
+	return vol.GetMountPoint()
+}
+
 // Create performs resource creation
 func (vol *VolumeResource) Create() error {
 	seelog.Debugf("Creating volume with name %s using driver %s", vol.Name, vol.VolumeConfig.Driver)
```
Comment (sharanyad) [inline agent/taskresource/volume/dockervolume.go]: nit: should this be `vol.DockerVolumeName`?

File: agent/taskresource/volume/dockervolume_test.go (hunk 34 above)
Comment (richardpen) [reply agent/taskresource/volume/dockervolume_test.go]: No, as you can see the times is zero. but I think we can just remove that `EXPECT`.

File: agent/taskresource/volume/dockervolume.go (hunk 35 above)
Comment (richardpen) [reply agent/taskresource/volume/dockervolume.go]: Yes, I will fix this in the other PR.
